cave/plot/algorithm_footprint.py

```diff
@@ -215,36 +278,177 @@ def get_2NN(x, X):
+                all = pd.concat([bad, good])
+            else:
+                all = pd.concat([good, bad])
+            len_longest = min(len(good), len(bad))
+            counts = all.groupby(all.columns.tolist(), as_index=False).size()  # count the occurance of values
+            if len(good) > 0: counts_g = good.groupby(good.columns.tolist(), as_index=False).size().unstack()  # in good
+            if len(bad) > 0: counts_b = bad.groupby(bad.columns.tolist(), as_index=False).size().unstack()  # and bad
+            for idx, coords in enumerate(all.values):  # individually plot the points
+                r, g, b = 0, 0, 0
+                if len(bad) > 0 and coords[0] in counts_b.index and coords[1] in counts_b.columns:  # determine red part
+                    # red part is the number of points on the same coordinate belonging to the bad group
+                    # divided by the number of all points on the same coordinate
+                    r = counts_b[coords[1]][coords[0]] / counts[coords[0]][coords[1]]
+                if len(good) > 0 and coords[0] in counts_g.index and coords[1] in counts_g.columns: # similar for green
+                    g = counts_g[coords[1]][coords[0]] / counts[coords[0]][coords[1]]
+                zorder = 1
+                alpha = 1
+                if len_longest < idx:  # if we plot points from the shorter list, increase zorder and use small alpha
+                    alpha = 0.375
+                    zorder=9999
+                plt.scatter(coords[0], coords[1], color=(r, g, b), s=15, zorder=zorder, alpha=alpha)
+            ax.set_ylabel('principal component 1')
+            ax.set_xlabel('principal component 2')
+            plt.tight_layout()
+            fig.savefig(out_fn)
+            plt.close(fig)
+            plots.append(out_fn)
+        return plots
+
+    def plot3d(self):
+        """ Plot 3d-version of the algorithm footprint from four different
+        angles. """
+        plots = []
+        for a in self.algorithms:
+            # Plot without clustering (for all insts)
+            out_fns = [os.path.join(self.output_dir, 'footprint_' +
+                      self.algo_names[a] + '_3d_{}.png'.format(i)) for i in range(4)]
+            self.logger.debug("Plot saved to '%s'", out_fns)
+            fig, ax = plt.subplots()
+            good_idx, bad_idx = self._get_good_bad(a)
+            good = np.array([self.features_3d[idx] for idx in good_idx])
+            bad = np.array([self.features_3d[idx] for idx in bad_idx])
+            # Plot 3d
+            fig = plt.figure()
+            ax = fig.add_subplot(111, projection='3d')
+            if len(good) > 0: ax.scatter(xs=good[:, 0], ys=good[:, 1],
+                                         zs=good[:, 2], color="green")
+            if len(bad) > 0: ax.scatter(xs=bad[:, 0], ys=bad[:, 1],
+                                        zs=bad[:, 2], color="red")
+            ax.set_ylabel('principal component 1', fontsize=12)
```

Comment:
Why is this specifically fixed to 12?
Doesn't this conflict with the mpl_style file?

Reply:
I think the fontsize for the 3d-plots should be an exception, since it doesn't work well:
![screenshot-2018-2-20 cave](https://user-images.githubusercontent.com/9417378/36478698-062be834-1706-11e8-83d2-21b97eea3f1b.png)

Reply:
I see.
Would that problem disappear with line breaks?
If not we could also shorten the Principal component to pc and explain what it means in the tooltip.